pykafka/rdkafka/_rd_kafkamodule.c

```diff
@@ -0,0 +1,1175 @@
+        self->rdk_topic_conf = NULL;  /* deallocated by rd_kafka_topic_new() */
+    Py_END_ALLOW_THREADS
+    if (! self->rdk_topic_handle) {
+        set_pykafka_error_from_code(rd_kafka_errno2err(errno), NULL);
+        return RdkHandle_start_fail(self, RdkHandle_stop);
+    }
+
+    if (RdkHandle_unlock(self)) return NULL;
+    Py_INCREF(Py_None);
+    return Py_None;
+}
+
+
+/**
+ * Producer type
+ */
+
+
+/* NB this doesn't check if RdkHandle_outq_len is zero, and generally assumes
+ * the wrapping python class will take care of ensuring any such preconditions
+ * for a clean termination */
+static void
+Producer_dealloc(PyObject *self)
+{
+    RdkHandle_dealloc(self, RdkHandle_stop);
+}
+
+
+/* Helper function for Producer_delivery_report_callback.  Because it is a
+ * non-python callback, it cannot raise exceptions.  We therefore try to log
+ * the traceback, then clear the exception */
+static void
+log_clear_exception(const char *msg)
+{
+    Py_XDECREF(PyObject_CallMethod(logger, "exception", "s", msg));
+    PyErr_Clear();
+}
+
+
+/* Helper function for Producer_delivery_report_callback: find an exception
+ * corresponding to `err`, and set that on `future`.  Returns -1 on failure,
+ * or 0 on success */
+static int
+Producer_delivery_report_set_exception(PyObject *future,
+                                       rd_kafka_resp_err_t err)
+{
+    int retval = 0;
+    PyObject *exc = NULL;
+
+    set_pykafka_error_from_code(err, &exc);
+    if (! exc) goto cleanup;
+    PyObject *res = PyObject_CallMethod(future, "set_exception", "O", exc);
+    if (! res) goto cleanup;
+    else Py_DECREF(res);
+
+cleanup:
+    if (PyErr_Occurred()) {
+        log_clear_exception("Couldn't pass error to Future.set_exception");
+        retval = -1;
+    }
+    Py_XDECREF(exc);
+    return retval;
+}
+
+/* Helper function to drop message from the keep-alive set `pending` */
+static int
+Producer_discard_from_pending(PyObject *pending, PyObject *message)
+{
+    int res = PySet_Discard(pending, message);
+    if (res == 1) return 0;
+    if (res == 0) set_pykafka_error("RdKafkaException",
+                                    "Message not found in pending messages!");
+    /* if (res == -1) error already set */
+    return -1;
+}
+
+
+/* Helper function to retrieve Message.delivery_future.  NB may return Py_None
+ * if user has already discarded future */
+PyObject *
+Producer_get_message_future(PyObject *message)
+{
+    PyObject *wref = NULL;
+    PyObject *future = NULL;
+
+    /* We expect message.delivery_future to be a weakref to a future */
+    wref = PyObject_GetAttrString(message, "delivery_future");
+    if (! wref) goto failed;
+    future = PyObject_CallObject(wref, NULL);
+    if (! future) goto failed;
+
+    Py_DECREF(wref);
+    return future;
+failed:
+    Py_XDECREF(wref);
+    Py_XDECREF(future);
+    return NULL;
+}
+
+
+static void
+Producer_delivery_report_callback(rd_kafka_t *rk,
```

Comment:
I guess this method will stay around, but it will not manipulate futures anymore, instead maybe pushing failures onto a queue.